File: Day326_2025/Q326_GFG.py
class DisjointSet:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n

    def find(self, node):
        if node == self.parent[node]:
            return node
        self.parent[node] = self.find(self.parent[node])  # Path compression
        return self.parent[node]

    def unionBySize(self, u, v):
        pu = self.find(u)
        pv = self.find(v)

        if pu == pv:
            return False

        if self.size[pu] > self.size[pv]:
            self.parent[pv] = pu
            self.size[pu] += self.size[pv]
        else:
            self.parent[pu] = pv
            self.size[pv] += self.size[pu]

        return True

    def countComp(self):
        return sum(1 for i in range(len(self.parent)) if self.parent[i] == i)


class Solution:
    def minConnect(self, V, edges):
        if len(edges) < V - 1:
            return -1
        
        dsu = DisjointSet(V)

        for u, v in edges:
            dsu.unionBySize(u, v)

        components = dsu.countComp()
        return components - 1

File: Day326_2025/Q326_GFG.java
class DisjointSet {
    int[] parent, size;

    public DisjointSet(int n) {
        parent = new int[n];
        size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    public int find(int node) {
        if (parent[node] == node) return node;
        parent[node] = find(parent[node]);  // Path compression
        return parent[node];
    }

    public boolean unionBySize(int u, int v) {
        int pu = find(u);
        int pv = find(v);

        if (pu == pv) return false;

        if (size[pu] > size[pv]) {
            parent[pv] = pu;
            size[pu] += size[pv];
        } else {
            parent[pu] = pv;
            size[pv] += size[pu];
        }
        return true;
    }

    public int countComp() {
        int c = 0;
        for (int i = 0; i < parent.length; i++) {
            if (parent[i] == i) c++;
        }
        return c;
    }
}

class Q326_GFG {
    public int minConnect(int V, int[][] edges) {
        int e = edges.length;

        if (e < V - 1) return -1;  // Not enough edges to connect all

        DisjointSet dsu = new DisjointSet(V);

        for (int[] edge : edges) {
            dsu.unionBySize(edge[0], edge[1]);
        }

        int components = dsu.countComp();
        return components - 1;  // Minimum operations
    }
}
